kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+-- @param opts table, requires `name`, optionally includes `retry_count`, `max_delay` and `batch_max_size`
+-- @return table: a Queue object.
+local function get_or_create_queue(queue_conf, handler, handler_conf)
+
+  local name = queue_conf.name
+
+  local queue = queues[name]
+  if queue then
+    queue:log(DEBUG, "queue exists")
+    -- We always use the latest configuration that we have seen for a queue and handler.
+    queue.handler_conf = handler_conf
+    return queue
+  end
+
+  queue = {
+    -- Queue parameters from the enqueue call
+    name = name,
+    handler = handler,
+    handler_conf = handler_conf,
+
+    -- semaphore to count the number of items on the queue and synchronize between enqueue and handler
+    semaphore = semaphore.new(),
+
+    -- `bytes_queued` holds the number of bytes on the queue.  It will be used only if string_capacity is set.
+    bytes_queued = 0,
+    -- `entries` holds the actual queue items.
+    entries = {},
+    -- Pointers into the table that holds the actual queued entries.  `front` points to the oldest, `back` points to
+    -- the newest entry.
+    front = 1,
+    back = 1,
+  }
+  for option, value in pairs(queue_conf) do
+    queue[option] = value
+  end
+
+  queue = setmetatable(queue, Queue_mt)
+
+  kong.timer:named_at(name, 0, function(_, q)
+    while q:count() > 0 do
+      q:log(DEBUG, "processing queue")
+      q:process_once()
+    end
+    q:log(DEBUG, "done processing queue")
+    queues[name] = nil
+  end, queue)
+
+  queues[name] = queue
+
+  queue:log(DEBUG, "queue created")
+
+  return queue
+end
+
+
+-------------------------------------------------------------------------------
+-- Log a message that includes the name of the queue for identification purposes
+-- @param self Queue
+-- @param level: log level
+-- @param formatstring: format string, will get the queue name and ": " prepended
+-- @param ...: formatter arguments
+function Queue:log(level, formatstring, ...)
+  local message = self.name .. ": " .. formatstring
+  if select('#', ...) > 0 then
+    return ngx.log(level, string.format(message, unpack({...})))
+  else
+    return ngx.log(level, message)
+  end
+end
+
+
+function Queue:count()
+  return self.back - self.front
+end
+
+
+-- Delete the frontmost entry from the queue and adjust the current utilization variables.
+function Queue:delete_frontmost_entry()
+  if self.string_capacity then
+    -- If string_capacity is set, reduce the currently queued byte count by the
+    self.bytes_queued = self.bytes_queued - #self.entries[self.front]
+  end
+  self.entries[self.front] = nil
+  self.front = self.front + 1
+  if self.front == self.back then
+    self.front = 1
+    self.back = 1
+  end
+end
+
+-- Process one batch of entries from the queue.  Returns truthy if entries were processed, falsy if there was an
+-- error or no items were on the queue to be processed.
+function Queue:process_once()
+  local ok, err = self.semaphore:wait(0)
```

Comment:
Why not look semaphore:count() to see how many resources are available to greedily pop items in batches?

Reply:
See above.